zkevm-circuits/src/evm_circuit/execution/codesize.rs

```diff
@@ -0,0 +1,110 @@
+use bus_mapping::evm::OpcodeId;
+use eth_types::Field;
+use halo2_proofs::plonk::Error;
+
+use crate::{
+    evm_circuit::{
+        step::ExecutionState,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition},
+            CachedRegion, Cell, RandomLinearCombination,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+
+use super::ExecutionGadget;
+
+#[derive(Clone, Debug)]
+pub(crate) struct CodesizeGadget<F> {
+    same_context: SameContextGadget<F>,
+    codesize: Cell<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for CodesizeGadget<F> {
+    const NAME: &'static str = "CODESIZE";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::CODESIZE;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        let code_hash = cb.curr.state.code_hash.clone();
+        let codesize = cb.bytecode_length(code_hash.expr());
+        let codesize_rlc = RandomLinearCombination::random_linear_combine_expr(
+            [codesize.expr()],
+            cb.power_of_randomness(),
+        );
+
+        cb.require_equal(
+            "Constraint: code size (bytecode table) == stack top",
+            codesize_rlc.expr(),
+            codesize.expr(),
+        );
```

Comment:
since the first argument of `random_linear_combine_expr` is a len-1 array, `random_linear_combine_expr` will become "codesize.expr() * 1".    So codesize_rlc is exact codesize. The "cb.require_equal" can never be false no matter what witness is assigned.

BTW i am a little afraid of the RLC, better to construct a test case with codesize larger than 256?

Reply:
Yes, you're correct. I missed this point. The PR was open for a while and I can see the mistake after reviewing the code myself now! Will fix this part